File: ex2.2.ts
const makePair = <T>(a: T, b: T): [T, T] => [a, b];

const getHead = <T>(pair: [T, T]) => pair[0];

const getTail = <T>(pair: [T, T]) => pair[1];

type Point = [number, number];

const makeSegment = (start: Point, end: Point) => makePair(start, end);
const getStartSegment = (segment: [Point, Point]) => getHead(segment);
const getEndSegment = (segment: [Point, Point]) => getTail(segment);

const makePoint = (x: number, y: number): Point => makePair(x, y);
const getPointX = (point: Point) => getHead(point);
const getPointY = (point: Point) => getTail(point);

const getMiddleSegment = (segment: [Point, Point]): Point => {
  let x = 0;
  let y = 0;

  const startPointX = getPointX(getStartSegment(segment));
  const startPointY = getPointY(getStartSegment(segment));
  const endPointX = getPointX(getEndSegment(segment));
  const endPointY = getPointY(getEndSegment(segment));
  if (startPointX > endPointX) {
    x = startPointX - endPointX;
  } else {
    x = endPointX - startPointX;
  }

  if (startPointY > endPointY) {
    y = startPointY - endPointY;
  } else {
    y = endPointY - startPointY;
  }

  return makePoint(x, y);
};

const printPoint = (point: Point) =>
  console.log(`x: ${getPointX(point)}, y: ${getPointY(point)}`);

const segment = makeSegment(makePoint(123, 2345), makePoint(2345, 546));

printPoint(getMiddleSegment(segment));

File: ex2.2.py
def make_pair(a, b):
    return [a, b]


def get_head(pair):
    return pair[0]


def get_tail(pair):
    return pair[1]


def make_segment(start, end):
    return make_pair(start, end)


def get_segment_start(segment):
    return get_head(segment)


def get_segment_end(segment):
    return get_tail(segment)


def make_point(x, y):
    return make_pair(x, y)


def get_point_x(point):
    return get_head(point)


def get_point_y(point):
    return get_tail(point)


def print_point(point):
    print(f"x: {get_head(point)} y:{get_tail(point)}")


def get_midpoint_segment(segment):
    segment_start_x = get_point_x(get_segment_start(segment))
    segment_end_x = get_point_x(get_segment_end(segment))
    segment_start_y = get_point_y(get_segment_start(segment))
    segment_end_y = get_point_y(get_segment_end(segment))
    x = 0
    y = 0

    if segment_start_x > segment_end_x:
        x = segment_start_x - segment_end_x
    else:
        x = segment_end_x - segment_start_x
    if segment_start_y > segment_end_y:
        y = segment_start_y - segment_end_y
    else:
        y = segment_end_y - segment_start_y

    return make_point(x, y)


segment = make_segment(make_point(1123, 123), make_point(123123, 1233))
print_point(get_midpoint_segment(segment))
